crates/polars-io/src/partition.rs

```diff
@@ -127,3 +127,69 @@ where
     }
     path
 }
+
+/// Creates an iterator of (hive partition path, DataFrame) pairs, e.g.:
+/// ("a=1/b=1", DataFrame)
+pub fn get_hive_partitions_iter<'a, S>(
+    df: &'a DataFrame,
+    partition_by: &'a [S],
+) -> PolarsResult<Box<dyn Iterator<Item = (String, DataFrame)> + 'a>>
+where
+    S: AsRef<str>,
+{
+    for x in partition_by.iter() {
+        if df.get_column_index(x.as_ref()).is_none() {
+            polars_bail!(ColumnNotFound: "{}", x.as_ref());
+        }
+    }
+
+    let get_hive_path_part = |df: &DataFrame| {
+        const CHAR_SET: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
+            .add(b'/')
+            .add(b'=')
+            .add(b':')
+            .add(b' ');
+
+        partition_by
+            .iter()
+            .map(|x| {
+                let s = df.column(x.as_ref()).unwrap().slice(0, 1);
+
+                format!(
+                    "{}={}",
+                    s.name(),
+                    percent_encoding::percent_encode(
+                        s.cast(&DataType::String)
+                            .unwrap()
+                            .str()
+                            .unwrap()
+                            .get(0)
+                            .unwrap_or("__HIVE_DEFAULT_PARTITION__")
+                            .as_bytes(),
+                        CHAR_SET
+                    )
+                )
+            })
+            .collect::<Vec<_>>()
+            .join("/")
+    };
+
+    let groups = df.group_by(partition_by)?;
+    let groups = groups.take_groups();
+
+    let out: Box<dyn Iterator<Item = (String, DataFrame)>> = match groups {
+        GroupsProxy::Idx(idx) => Box::new(idx.into_iter().map(move |(_, group)| {
+            let part_df =
```

Comment:
I think we should iterate over `DataFrame`s of a certain size. So that we don't write a single file per folder, but for large partitions many smaller parquet files.

I am not entirely sure how other tools determine the size of the parquet. We could split by `n_rows` where we use the `estimated_size` as hint?

Reply:
I set it to 1 million rows per file for now

Reply:
If you know the schema at this point (or rather, the number of cols) it's better to target a given number of _elements_ (rows x cols), as "rows" by itself is not a useful metric.

1 million rows with 1 col is a full three orders of magnitude removed from 1 million rows with 1000 cols 😆

Somewhere between 10-25 million elements is probably going to be a more consistent target 🤔

Reply:
I see, I've changed it slice the df into chunks of a target size